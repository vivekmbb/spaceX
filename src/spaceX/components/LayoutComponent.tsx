import React from "react";
import { Box, Grid, ResponsiveContext, Text } from "grommet";
import { getResults } from "../services/api";
import { FilterSection } from "./FilterSection";
import { MainPageComponent } from "./MainPageComponent";
import { useHistory } from "react-router-dom";

export const LayoutComponent = () => {
  const [selectYear, setSelectYear] = React.useState();
  const [selectLaunch, setSelectLaunch] = React.useState();
  const [selectLanding, setSelectLanding] = React.useState();

  const history = useHistory();

  const props = {
    selectYear,
    setSelectYear,
    selectLaunch,
    setSelectLaunch,
    selectLanding,
    setSelectLanding
  };

  const [results, setResults] = React.useState<any>();
  const size = React.useContext(ResponsiveContext);

  const columns = size === "large" || size === "medium" ? ["1/4", "3/4"] : [];
  const rows = size === "large" || size === "medium" ? ["40px", "flex", "80px"] : ["40px", "500px", "large", "80px"];
  const areas =
    size === "large" || size === "medium"
      ? [
          { name: "header", start: [0, 0], end: [1, 0] },
          { name: "filter", start: [0, 1], end: [0, 1] },
          { name: "main", start: [1, 1], end: [1, 1] },
          { name: "footer", start: [0, 2], end: [1, 2] }
        ]
      : [
          { name: "header", start: [0, 0], end: [0, 0] },
          { name: "filter", start: [0, 1], end: [0, 1] },
          { name: "main", start: [0, 2], end: [0, 2] },
          { name: "footer", start: [0, 3], end: [0, 3] }
        ];

  React.useEffect(() => {
    getData();
    return () => {
      history.push("/");
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectYear, selectLaunch, selectLanding]);

  const getData = async () => {
    try {
      const data = await getResults(constructUriWithParams());
      setResults(data.data);
    } catch (e) {
      alert("Error Getting data");
    }
  };

  const constructUriWithParams = () => {
    const defaultParam = `limit=100`;
    const year = selectYear ? `&launch_year=${selectYear}` : "";
    const launch = selectLaunch ? `&launch_success=${selectLaunch === "True" ? true : false}` : "";
    const landing = selectLanding ? `&land_success=${selectLanding === "True" ? true : false}` : "";
    const finalUri = `https://api.spaceXdata.com/v3/launches?limit=100${year}${launch}${landing}`;
    history.push(`${defaultParam}${year}${launch}${landing}`);
    return finalUri;
  };

  return (
    <Grid rows={rows} columns={columns} fill={"vertical"} areas={areas}>
      <Box gridArea="header" background="light-3">
        <Text size="large" weight="bold" margin={{ left: "xsmall", top: "xsmall" }}>
          SpaceX Launch Programs
        </Text>
      </Box>
      <Box gridArea="filter" background="light-3" pad={size === "small" ? "medium" : "small"}>
        <FilterSection props={props} />
      </Box>
      <Box
        gridArea="main"
        background="light-3"
        pad={size === "small" ? "medium" : "small"}
        margin={{ right: "small" }}
        direction="column"
        overflow="auto"
      >
        {results && <MainPageComponent result={results} />}
      </Box>
      <Box gridArea="footer" background="light-3" align="center">
        <Text size="large" weight="bold" margin="medium">
          Developed By : Vivek
        </Text>
      </Box>
    </Grid>
  );
};
