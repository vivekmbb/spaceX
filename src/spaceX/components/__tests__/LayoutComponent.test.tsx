// tslint:disable: no-submodule-imports
import "@testing-library/jest-dom/extend-expect";
import { Box } from "grommet";
import React from "react";
import renderer from "react-test-renderer";
import { LayoutComponent } from "../LayoutComponent";

describe("Layout Component Information Component", () => {
  it("Layout component", () => {
    const Wrapper = (props: any) => <Box pad="medium">{props.children}</Box>;
    const component = renderer
      .create(
        <Wrapper>
          <LayoutComponent />
        </Wrapper>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
