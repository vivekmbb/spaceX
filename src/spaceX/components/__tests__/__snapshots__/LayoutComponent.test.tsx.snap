// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Component Information Component Layout component 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 chkvLT"
>
  <div
    className="StyledGrid-sc-1wofa1l-0 hPSYfF"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 RcNfk"
    >
      <span
        className="StyledText-sc-1sadyjn-0 eUnrYS"
      >
        SpaceX Launch Programs
      </span>
    </div>
    <div
      className="StyledBox-sc-13pk1d4-0 dHfVzm"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 irFSzw"
      >
        <span
          className="StyledText-sc-1sadyjn-0 eeaybW"
        >
          Filters
        </span>
        <div
          className="StyledBox-sc-13pk1d4-0 fPUIqt"
        >
          <span
            className="StyledText-sc-1sadyjn-0 toSoo"
          >
            Launch Year
          </span>
        </div>
        <div
          className="StyledBox-sc-13pk1d4-0 fCcErq"
          data-testid="button section"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2006
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2007
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2008
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2009
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2010
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2011
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2012
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2013
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2014
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2015
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2016
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2017
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2018
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2019
            </div>
          </div>
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              2020
            </div>
          </div>
        </div>
        <div
          className="StyledBox-sc-13pk1d4-0 dpWyCI"
        >
          <span
            className="StyledText-sc-1sadyjn-0 toSoo"
          >
            Successful Launch
          </span>
        </div>
        <div
          className="StyledBox-sc-13pk1d4-0 fCcErq"
          data-testid="button section"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              True
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              False
            </div>
          </div>
        </div>
        <div
          className="StyledBox-sc-13pk1d4-0 dpWyCI"
        >
          <span
            className="StyledText-sc-1sadyjn-0 toSoo"
          >
            Successful Landing
          </span>
        </div>
        <div
          className="StyledBox-sc-13pk1d4-0 fCcErq"
          data-testid="button section"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 iSmarO"
          >
            <div
              className="StyledBox-sc-13pk1d4-0 IjQKZ"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              True
            </div>
            <div
              className="StyledBox__StyledBoxGap-sc-13pk1d4-1 kxsFh"
            />
            <div
              className="StyledBox-sc-13pk1d4-0 kiMtHO"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            >
              False
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="StyledBox-sc-13pk1d4-0 nJbPu"
    />
    <div
      className="StyledBox-sc-13pk1d4-0 bDHRrK"
    >
      <span
        className="StyledText-sc-1sadyjn-0 dpybyU"
      >
        Developed By : Vivek
      </span>
    </div>
  </div>
</div>
`;
